pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go

```diff
@@ -0,0 +1,142 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+)
+
+type ServiceExportController struct {
+	client        versioned.Interface
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	environment *model.Environment
+
+	// We use this flag to short-circuit the logic and stop the controller if the CRD does not exist (or is deleted)
+	crdExists bool
+}
+
+func NewServiceExportController(kubeClient kube.Client, environment *model.Environment) (*ServiceExportController, error) {
+	serviceExportController := &ServiceExportController{
+		client:        kubeClient.MCSApis(),
+		serviceClient: kubeClient.Kube().CoreV1(),
+		queue:         queue.NewQueue(time.Second),
+		environment:   environment,
+		crdExists:     true,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
+	serviceExportController.serviceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { serviceExportController.onServiceAdd(obj) },
+		DeleteFunc: func(obj interface{}) { serviceExportController.onServiceDelete(obj) },
+	})
+
+	return serviceExportController, nil
+}
+
+func (sc *ServiceExportController) onServiceAdd(obj interface{}) {
+	sc.queue.Push(func() error {
+		serviceObj, err := convertToService(obj)
+		if err != nil {
+			return err
+		}
+		return sc.handleNewService(serviceObj)
+	})
+}
+
+func (sc *ServiceExportController) onServiceDelete(obj interface{}) {
+	sc.queue.Push(func() error {
+		serviceObj, err := convertToService(obj)
+		if err != nil {
+			return err
+		}
+		return sc.handleDeletedService(serviceObj)
+	})
+}
+
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) handleNewService(obj *v1.Service) error {
+	if !sc.crdExists {
+		return nil
+	}
+	if sc.environment.IsServiceClusterLocal(obj) {
+		return nil // don't do anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) handleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = service.Namespace
+	serviceExport.Name = service.Name
```

Comment:
Do you want to add an owner reference to this?  This would also eliminate the need for handling deletions.

Reply:
That is actually something that is explicitly called out by the [K8s MCS Spec](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#objectreference-in-serviceexportspec-to-directly-map-to-a-service), and was decided against. Instead, they chose to rely entirely on the name mapping. I tried to make this implementation match the spec as closely as possible.